specs/merkle-tree.md

```diff
@@ -0,0 +1,166 @@
+
+## Assumptions
+
+### Executor Behavior Regarding Invalid Merkle Trees
+The Merkle tree generation utility provided here is used to generate Merkle trees for the official Sphinx Foundry plugin, but there is not hard requirement that it be used to generate all Merkle trees. It is entirely possible that some third-party could implement their own plugin and with some alternative generation logic.
+
+Therefore, it is possible to use alternative generation logic to constuct Merkle trees that fit the high level [architecture](#merkle-tree-architecture) of the Sphinx Merkle tree, but that are not executable and are invalid according to the [invariants](#high-level-merkle-tree-invariants). For example, one could create a tree that contains multiple leaves with the same `index` and `chainId` fields. In this case, it would only be possible to execute one of these leaves. As a result, it would not be possible to fully execute this Merkle tree and furthermore it would be ambiguous exactly which leaf should be executed if one were to attempt to execute the tree.
+
+It is also possible to construct Merkle trees that are invalid according to the invariants defined above, but are still executable on-chain. For example, one could construct a Merkle tree with a different method of determining the position of leaves within the tree (such as reverse or a random order), but that still have valid `index` and `chainId` fields for every leaf. It would be possible to execute a tree like this as long as the leaves are submitted in the correct order according to their `index` field. However, this tree would not be valid according to the invariants defined above.
+
+We assume that executors like the Sphinx DevOps platform will use the invariants defined in this document to determine whether or not a given Merkle tree is valid. Furthermore, we assume they will refuse to execute Merkle trees that are not valid according to the invariants defined in this document even if it may theoretically be possible to partially or fully execute the Merkle tree.
+
+### Input Data
+Merkle trees generated with this logic are only valid to the extent that the data used to generate them is valid. For example, you could input a `gas` value for a transaction that is not sufficient which would result in the transaction failing. We assume that the input data satisfies the [on-chain conditions in the corresponding SphinxModuleProxy contract(s)](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/specs/sphinx-module-proxy.md#function-level-invariants).
```

Comment:
> We assume that the input data satisfies the [on-chain conditions in the corresponding SphinxModuleProxy contract(s)](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/specs/sphinx-module-proxy.md#function-level-invariants).

Not sure what you mean by "on-chain conditions". Did you mean to say "assumptions" instead?

Reply:
Things like the addresses of the safe and module proxies, the nonce, etc. I will update this.